### ISSUE TITLE
Guidance Interpreting 2D and 3D Keypoints

### ISSUE THREAD

**Issue by async-alchemy** [rolpotamias/WiLoR#8]
Hi! I loved using the model; phenomenal performance.

I’m hoping to utilize the 2D and 3D predicted keypoints from the model in a downstream task. However, I’m wondering how to interpret them? I see they’re all float values between (-1, 1)…does that refer to the percent distance from the center of the bounding box to the edge?

is there also a mapping for keypoints idx -> body part i can reference to understand what each keypoint is identifying?

I’m hoping to translate the keypoints into pixel values on the input image. 

Thanks in advance!

**Comment by rolpotamias** [comment]
Hi there. 

The values that you refer to are the normalized image coordinates for each crop. I have updated the `demo.py` and now you can directly get the 2D hand coordinates on the full image. Please check [here](https://github.com/rolpotamias/WiLoR/blob/main/demo.py#L97) 

Regarding the keypoint indices,  we use the classic MANO ordering: 
![image](https://github.com/user-attachments/assets/57b3c51c-5bd8-49fb-b4cb-b997b0dbb115)

**Comment by georgeYanch** [comment]
@async-alchemy Here's my attempt to transform 2d keypoints from crop to image coordinates:

[here ](https://github.com/rolpotamias/WiLoR/blob/main/wilor/datasets/vitdet_dataset.py#L74) you can get affine transformation matrix and add it to returned dict:

```
        item = {
                    'img': img_patch,
                    'personid': int(self.personid[idx]),
                }
        item['box_center'] = self.center[idx].copy()
        item['box_size'] = bbox_size
        item['img_size'] = 1.0 * np.array([cvimg.shape[1], cvimg.shape[0]])
        item['right'] = self.right[idx].copy()
        item['trans'] = trans # Here
        return item
```

 Then you can access it from generated batch:

```
        multiplier    = (2*batch['right']-1)
        pred_cam      = out['pred_cam']
        pred_cam[:,1] = multiplier*pred_cam[:,1]
        box_center    = batch["box_center"].float()
        box_size      = batch["box_size"].float()
        img_size      = batch["img_size"].float()
        trans         = batch["trans"].detach().cpu().numpy()[0]
```

After that

```
        joints2d = out['pred_keypoints_2d'][n].detach().cpu().numpy()
        joints2d[:, [0, 1]] *= 256
        joints2d[:, [0, 1]] += 128 
        joints2d = restore_keypoints_to_full_image(joints2d, trans)
```

Function for applying inverse affine matrix:

```
        def restore_keypoints_to_full_image(keypoints: np.array, trans: np.array) -> np.array:
            keypoints_h = np.hstack([keypoints, np.ones((keypoints.shape[0], 1))])
            trans_inv = cv2.invertAffineTransform(trans)
            restored_keypoints = keypoints_h @ trans_inv.T
            return restored_keypoints
```

And don't forget to subtract x-coords from width of full image:

`          plt.scatter(orig_image_width - joints2d[:, 0], joints2d[:, 1])`

I believe there's a simpler way to do it, but it works, I guess

**Comment by xiyangyang99** [comment]
> Hi there.
> 
> The values that you refer to are the normalized image coordinates for each crop. I have updated the `demo.py` and now you can directly get the 2D hand coordinates on the full image. Please check [here](https://github.com/rolpotamias/WiLoR/blob/main/demo.py#L97)
> 
> Regarding the keypoint indices, we use the classic MANO ordering: ![image](https://private-user-images.githubusercontent.com/36522591/374754403-57b3c51c-5bd8-49fb-b4cb-b997b0dbb115.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAyODMxODIsIm5iZiI6MTczMDI4Mjg4MiwicGF0aCI6Ii8zNjUyMjU5MS8zNzQ3NTQ0MDMtNTdiM2M1MWMtNWJkOC00OWZiLWI0Y2ItYjk5N2IwZGJiMTE1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDMwVDEwMDgwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE1YWU4OTUzMDNmMzg4MjQ2YWE0MmM3YzFiNjU2MTBlZWNmNjc3ZmQ5NzQwMWJmMmE0OWNmMTgzMzk0YWY0ODkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0._HQP2yjwsfGZlhUMTro_FelQ0a2ekPX7fG6r7kPalWY)

Hello author, may I ask how to obtain the predicted 3D coordinates of the hand in the form of θ, β, and π.
